[Filebeat][httpjson] Convert httpjson input to a v2 input

marc-gr opened the pull request:
## What does this PR do?

Converts httpjson input to the new v2 stateless input.

- Split old input in several new objects for easier understanding
- Refactored go tests to simplify them and make them work with the new input

## Why is it important?

Several modules rely on httpjson input, and would benefit a lot from the new v2 cursor feature that keeps persistence of a cursor between restarts. 

This change moves httpjson input to new v2 input, preparing to be able to implement the cursor persistence state in a following change.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
~- [] I have made corresponding changes to the documentation~
~- [] I have made corresponding change to the default configuration files~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

Relates to #19486

elasticmachine commented:
Pinging @elastic/siem (Team:SIEM)

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > There is a new build on-going so the previous on-going builds have been aborted.
  


<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-20226/detail/PR-20226/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-20226/detail/PR-20226/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-20226/detail/PR-20226/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-20226/detail/PR-20226/1//artifacts) 

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #20226 opened]`

* Reason: `Aborted from #2`

* Start Time: 2020-07-24T09:25:14.694+0000

* Duration: 4 min 46 sec


* Commit: 4bb2e72df53e51db06c9a1e1065c44698f030b4d




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors

  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  * **Name**: `Error signal`
    * Description: tar: step failled with error null
    
    * Duration: 0 min 0 sec
    
    * Start Time: 2020-07-24T09:28:50.389+0000
    
    * [log](https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-20226/runs/1/steps/85/log)
    
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-07-24T09:28:52.158Z] Stage "Elastic Agent x-pack Windows" skipped due to earlier failure(s)
[2020-07-24T09:28:52.178Z] Stage "Elastic Agent Mac OS X" skipped due to earlier failure(s)
[2020-07-24T09:28:52.190Z] Stage "Filebeat oss" skipped due to earlier failure(s)
[2020-07-24T09:28:52.201Z] Stage "Filebeat x-pack" skipped due to earlier failure(s)
[2020-07-24T09:28:52.213Z] Stage "Filebeat Mac OS X" skipped due to earlier failure(s)
[2020-07-24T09:28:52.224Z] Stage "Filebeat x-pack Mac OS X" skipped due to earlier failure(s)
[2020-07-24T09:28:52.241Z] Stage "Filebeat Windows" skipped due to earlier failure(s)
[2020-07-24T09:28:52.252Z] Stage "Filebeat x-pack Windows" skipped due to earlier failure(s)
[2020-07-24T09:28:52.265Z] Stage "Heartbeat" skipped due to earlier failure(s)
[2020-07-24T09:28:52.271Z] Stage "Auditbeat oss Linux" skipped due to earlier failure(s)
[2020-07-24T09:28:52.283Z] Stage "Auditbeat crosscompile" skipped due to earlier failure(s)
[2020-07-24T09:28:52.295Z] Stage "Auditbeat oss Mac OS X" skipped due to earlier failure(s)
[2020-07-24T09:28:52.308Z] Stage "Auditbeat oss Windows" skipped due to earlier failure(s)
[2020-07-24T09:28:52.321Z] Stage "Auditbeat x-pack" skipped due to earlier failure(s)
[2020-07-24T09:28:52.334Z] Stage "Auditbeat x-pack Mac OS X" skipped due to earlier failure(s)
[2020-07-24T09:28:52.347Z] Stage "Auditbeat x-pack Windows" skipped due to earlier failure(s)
[2020-07-24T09:28:52.359Z] Stage "Libbeat" skipped due to earlier failure(s)
[2020-07-24T09:28:52.365Z] Stage "Libbeat x-pack" skipped due to earlier failure(s)
[2020-07-24T09:28:52.378Z] Stage "Metricbeat OSS Unit tests" skipped due to earlier failure(s)
[2020-07-24T09:28:52.390Z] Stage "Metricbeat OSS Go Integration tests" skipped due to earlier failure(s)
[2020-07-24T09:28:52.402Z] Stage "Metricbeat OSS Python Integration tests" skipped due to earlier failure(s)
[2020-07-24T09:28:52.415Z] Stage "Metricbeat x-pack" skipped due to earlier failure(s)
[2020-07-24T09:28:52.421Z] Stage "Metricbeat crosscompile" skipped due to earlier failure(s)
[2020-07-24T09:28:52.434Z] Stage "Metricbeat Mac OS X" skipped due to earlier failure(s)
[2020-07-24T09:28:52.448Z] Stage "Metricbeat x-pack Mac OS X" skipped due to earlier failure(s)
[2020-07-24T09:28:52.461Z] Stage "Metricbeat Windows" skipped due to earlier failure(s)
[2020-07-24T09:28:52.475Z] Stage "Metricbeat x-pack Windows" skipped due to earlier failure(s)
[2020-07-24T09:28:52.488Z] Stage "Packetbeat OSS" skipped due to earlier failure(s)
[2020-07-24T09:28:52.494Z] Stage "dockerlogbeat" skipped due to earlier failure(s)
[2020-07-24T09:28:52.500Z] Stage "Winlogbeat" skipped due to earlier failure(s)
[2020-07-24T09:28:52.506Z] Stage "Winlogbeat Windows x-pack" skipped due to earlier failure(s)
[2020-07-24T09:28:52.518Z] Stage "Functionbeat" skipped due to earlier failure(s)
[2020-07-24T09:28:52.525Z] Stage "Journalbeat" skipped due to earlier failure(s)
[2020-07-24T09:28:52.531Z] Stage "Generators" skipped due to earlier failure(s)
[2020-07-24T09:28:52.537Z] Stage "Kubernetes" skipped due to earlier failure(s)
[2020-07-24T09:28:52.768Z] Stage "Heartbeat" skipped due to earlier failure(s)
[2020-07-24T09:28:52.777Z] Stage "Libbeat" skipped due to earlier failure(s)
[2020-07-24T09:28:52.787Z] Stage "Metricbeat x-pack" skipped due to earlier failure(s)
[2020-07-24T09:28:52.796Z] Stage "Packetbeat OSS" skipped due to earlier failure(s)
[2020-07-24T09:28:52.807Z] Stage "dockerlogbeat" skipped due to earlier failure(s)
[2020-07-24T09:28:52.816Z] Stage "Winlogbeat" skipped due to earlier failure(s)
[2020-07-24T09:28:52.826Z] Stage "Functionbeat" skipped due to earlier failure(s)
[2020-07-24T09:28:52.836Z] Stage "Journalbeat" skipped due to earlier failure(s)
[2020-07-24T09:28:52.846Z] Stage "Generators" skipped due to earlier failure(s)
[2020-07-24T09:28:54.132Z] Failed in branch Elastic Agent x-pack
[2020-07-24T09:28:54.141Z] Failed in branch Elastic Agent x-pack Windows
[2020-07-24T09:28:54.150Z] Failed in branch Elastic Agent Mac OS X
[2020-07-24T09:28:54.159Z] Failed in branch Filebeat oss
[2020-07-24T09:28:54.170Z] Failed in branch Filebeat x-pack
[2020-07-24T09:28:54.179Z] Failed in branch Filebeat Mac OS X
[2020-07-24T09:28:54.188Z] Failed in branch Filebeat x-pack Mac OS X
[2020-07-24T09:28:54.197Z] Failed in branch Filebeat Windows
[2020-07-24T09:28:54.206Z] Failed in branch Filebeat x-pack Windows
[2020-07-24T09:28:54.215Z] Failed in branch Auditbeat oss Linux
[2020-07-24T09:28:54.224Z] Failed in branch Auditbeat crosscompile
[2020-07-24T09:28:54.233Z] Failed in branch Auditbeat oss Mac OS X
[2020-07-24T09:28:54.241Z] Failed in branch Auditbeat oss Windows
[2020-07-24T09:28:54.250Z] Failed in branch Auditbeat x-pack
[2020-07-24T09:28:54.259Z] Failed in branch Auditbeat x-pack Mac OS X
[2020-07-24T09:28:54.269Z] Failed in branch Auditbeat x-pack Windows
[2020-07-24T09:28:54.279Z] Failed in branch Libbeat x-pack
[2020-07-24T09:28:54.288Z] Failed in branch Metricbeat OSS Unit tests
[2020-07-24T09:28:54.297Z] Failed in branch Metricbeat OSS Go Integration tests
[2020-07-24T09:28:54.306Z] Failed in branch Metricbeat OSS Python Integration tests
[2020-07-24T09:28:54.315Z] Failed in branch Metricbeat crosscompile
[2020-07-24T09:28:54.324Z] Failed in branch Metricbeat Mac OS X
[2020-07-24T09:28:54.333Z] Failed in branch Metricbeat x-pack Mac OS X
[2020-07-24T09:28:54.342Z] Failed in branch Metricbeat Windows
[2020-07-24T09:28:54.351Z] Failed in branch Metricbeat x-pack Windows
[2020-07-24T09:28:54.360Z] Failed in branch Winlogbeat Windows x-pack
[2020-07-24T09:28:54.370Z] Failed in branch Kubernetes
[2020-07-24T09:28:54.944Z] Stage "Heartbeat" skipped due to earlier failure(s)
[2020-07-24T09:28:54.955Z] Stage "Libbeat" skipped due to earlier failure(s)
[2020-07-24T09:28:54.969Z] Stage "Metricbeat x-pack" skipped due to earlier failure(s)
[2020-07-24T09:28:54.981Z] Stage "Packetbeat OSS" skipped due to earlier failure(s)
[2020-07-24T09:28:54.990Z] Stage "Winlogbeat" skipped due to earlier failure(s)
[2020-07-24T09:28:55.000Z] Stage "Functionbeat" skipped due to earlier failure(s)
[2020-07-24T09:28:55.010Z] Stage "Generators" skipped due to earlier failure(s)
[2020-07-24T09:28:55.412Z] Failed in branch dockerlogbeat
[2020-07-24T09:28:55.421Z] Failed in branch Journalbeat
[2020-07-24T09:28:55.905Z] Stage "Heartbeat" skipped due to earlier failure(s)
[2020-07-24T09:28:55.915Z] Stage "Libbeat" skipped due to earlier failure(s)
[2020-07-24T09:28:55.924Z] Stage "Packetbeat OSS" skipped due to earlier failure(s)
[2020-07-24T09:28:55.934Z] Stage "Functionbeat" skipped due to earlier failure(s)
[2020-07-24T09:28:55.944Z] Stage "Generators" skipped due to earlier failure(s)
[2020-07-24T09:28:56.051Z] Failed in branch Metricbeat x-pack
[2020-07-24T09:28:56.061Z] Failed in branch Winlogbeat
[2020-07-24T09:28:56.552Z] Failed in branch Heartbeat
[2020-07-24T09:28:56.562Z] Failed in branch Libbeat
[2020-07-24T09:28:56.573Z] Failed in branch Packetbeat OSS
[2020-07-24T09:28:56.585Z] Failed in branch Functionbeat
[2020-07-24T09:28:56.586Z] Stage "Generators" skipped due to earlier failure(s)
[2020-07-24T09:28:56.793Z] Failed in branch Generators
[2020-07-24T09:28:58.941Z] Running on Jenkins in /var/lib/jenkins/workspace/Beats_beats_PR-20226
[2020-07-24T09:28:59.235Z] [INFO] getVaultSecret: Getting secrets
[2020-07-24T09:28:59.354Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-07-24T09:29:00.525Z] + chmod 755 generate-build-data.sh
[2020-07-24T09:29:00.525Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-20226/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-20226/runs/1 ABORTED 225561
[2020-07-24T09:29:00.525Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-20226/runs/1/steps/?limit=10000 -o steps-info.json
[2020-07-24T09:29:00.776Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-20226/runs/1/tests/?status=FAILED -o tests-errors.json

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

andrewkroh commented on x-pack/filebeat/input/httpjson/input.go:
```diff
@@ -5,596 +5,152 @@
 
-	if mm != nil && in.config.DateCursor.IsEnabled() {
-		in.advanceCursor(common.MapStr(mm))
+	_, err = net.DialTimeout("tcp", fmt.Sprintf("%s:%s", url.Hostname(), port), time.Second)
```
Should this use `net.JoinHostPort` to ensure IPv6 addresses are boxed? I think url.Hostname() strips brackets.

urso commented on x-pack/filebeat/input/httpjson/input.go:
```diff
@@ -5,596 +5,152 @@
-		workerCtx:    workerCtx,
-		workerCancel: workerCancel,
-	}
+	in := &httpJSONInput{}
```
It looks the requester does too much. It is a 'source', unpacks from the cursor, is a client, and executes the request. Plus the requester gets only partially initialized here. The source is used to allow to generate the key in the key value store, and to pass the input the address where to collect from. No need to confiure an http client yet. Configure should focus on unpacking the config, do not create helper objects that are required for data collection yet.

Following the code your source is the URL only. Maybe you want to consider to add other metadata to the source as well, e.g. parameters, username (in case services return different information based on the login).

Pass the config to the httpJSONInput, not your source. The requester can be created in the "Run" method.

marc-gr replied:
I agree, tried to separate pagination, cursor and retry logic out of it, lmk if it still confusing.

urso commented on x-pack/filebeat/input/httpjson/requester.go:
```diff
@@ -0,0 +1,464 @@
+	config      config
+	client      *http.Client
+	cursorValue string
+}
```
this type does a lot. It implements rate limiting, request execution, response parsing+event creation, and publishing. It looks like it is very difficult to test. Can we split some functionality into a dedicated type and introduce unit testing? At least for rate limiting consider to use/introduce a third party library.

marc-gr replied:
For the rate limit bit, it was not in the scope of this change, but it makes sense, will open a new issue to make it separately if that makes sense.

marc-gr commented on x-pack/filebeat/input/httpjson/requester.go:
```diff
@@ -0,0 +1,464 @@
+	}
+}
+
+func (l *retryLogger) Printf(s string, args ...interface{}) {
```
I think you were referring to implement the `LeveledLogger`, I did so, let me know if you meant something different.

marc-gr commented on x-pack/filebeat/tests/system/test_httpjson.py:
```diff
@@ -0,0 +1,611 @@
+from filebeat import BaseTest
+
+
+class Test(BaseTest):
```
I had some issues trying to spawn the cursor input manager in unit tests, that was the reason why I moved the integration tests to python. Since your following comments made a lot of sense, and I implemented the stateless input instead as a first step, I moved them back to go now, since the issues I faced with the cursor input seemed to not apply with that one.

marc-gr commented on x-pack/filebeat/input/httpjson/input.go:
```diff
@@ -5,596 +5,152 @@
+			StateStore: store,
+			Type:       inputName,
+			Configure:  configure,
+		},
```
Make sense, did so 👍

marc-gr commented:
Following @urso advice I pulled back and implemented stateless input instead. Will let the work of moving to a custom input manager to be able to use also the cursor input for a following pull request.

At the same time did some refactoring to simplify input/requester logic. More improvements can be done but I think would be nice to save them for the following refactor of the input that we plan to do next.

Updated the description to make it describe what it currently does.

marc-gr commented on x-pack/filebeat/input/httpjson/config_oauth.go:
```diff
@@ -66,8 +66,8 @@ func (o *OAuth2) IsEnabled() bool {
 // Client wraps the given http.Client and returns a new one that will use the oauth authentication.
-func (o *OAuth2) Client(ctx context.Context, client *http.Client) (*http.Client, error) {
-	ctx = context.WithValue(ctx, oauth2.HTTPClient, client)
+func (o *OAuth2) Client(client *http.Client) (*http.Client, error) {
```
Since was creating the client at configure step, there was no context yet, this got solved by doing as you suggested and creating it at `Run`

urso commented:
> This allowed to go back to go tests since the issues I faced when writing them for the cursor input do not apply for this one.

Maybe let's discuss this offline, but it would be nice to learn where you did struggle with testing. In libbeat we have helpers for mocking beat.Pipeline, and beat.Client. The `libbeat/statestore/storetest` package provides helpers to construct an in memory store that can be checked (state is stored in a `map[string]interface{}`). For example check out `filebeat/input/v2/input-cursor/store_test.go`. We got many expensive python testing for functionality that should be unit or integration tested only. If you find it difficult with existing helpers to write test, let's talk about it to find what is missing. All in all I'd like to encourage devs to write go based tests.

urso commented on x-pack/filebeat/input/httpjson/date_cursor.go:
```diff
@@ -0,0 +1,111 @@
+
+	url.RawQuery = q.Encode()
+
+	return url.String()
```
why return the string instead of the already initialized URL struct?

urso commented on x-pack/filebeat/input/httpjson/httpjson_test.go:
```diff
@@ -6,783 +6,409 @@ package httpjson
+			teardown: func(i interface{}) {
+				server := i.(*httptest.Server)
+				server.Close()
+			},
```
do we ever return some other type but an http server?

teardown as is can be removed. Better pass `testing.T` to setup. the `close` can be registered with the test via `t.Cleanup`.

For this tests we end up with 2 active servers. how about only providing a `setupServer` method? then `ssl`, `setup`, and `teardown` can be replaced by a single replacable function.

urso commented on x-pack/filebeat/input/httpjson/httpjson_test.go:
```diff
@@ -6,783 +6,409 @@ package httpjson
-			t.Fatal(err)
-		}
-	})
+type publisher struct {
```
The libbeat/publisher/testing might already contain similar helpers.

urso commented on x-pack/filebeat/input/httpjson/input.go:
```diff
@@ -81,532 +47,146 @@ func newRetryLogger() *retryLogger {
 				}
 			}
 		}
 	}
```
This loop is quite a common pattern. Better use `timed.Periodic(stdCtx, in.Config.Interval, func() { ... })`. Using this pattern to check for cancellation in a loop one should add an extra select on the done channel only in order to confirm done has not been closed before doing the actual select statement.

Periodic is defined here: https://github.com/elastic/go-concert/blob/master/timed/timed.go#L57

urso commented on x-pack/filebeat/input/httpjson/input.go:
```diff
@@ -81,532 +47,146 @@ func newRetryLogger() *retryLogger {
+func newHTTPJSONInput(config config) (*httpJSONInput, error) {
+	if err := config.Validate(); err != nil {
+		return nil, err
 	}
```
note: Validate is already called by Unpack. But I like it to be called here again, as the constructor should be independent of `configure`.

urso commented on x-pack/filebeat/input/httpjson/input.go:
```diff
@@ -81,532 +48,140 @@ func newRetryLogger() *retryLogger {
+			log.Info("Process another repeated request.")
+			err = requester.processHTTPRequest(stdCtx, publisher)
+			if err != nil {
+				periodCancel()
```
the Periodic function will support error returns to stop the loop once https://github.com/elastic/beats/pull/20590 is merged.

urso commented on x-pack/filebeat/input/httpjson/input.go:
```diff
@@ -81,532 +49,150 @@ func newRetryLogger() *retryLogger {
+	url, err := url.Parse(in.config.URL)
+	if err != nil {
+		return err
+	}
```
This is clearly a config parsing step and should therefore be done in the configure or newXInput function. Catch faulty configurations early. For URLs it would be nice to have a custom type that implements `Unpack(in string)`. All in all it would be nice to have a 'settings' package with reusable types that support Unpacking and Validate.

urso commented on x-pack/filebeat/input/httpjson/input.go:
```diff
@@ -81,532 +49,150 @@ func newRetryLogger() *retryLogger {
-		in.nextCursorValue = time.Now().UTC().Format(in.config.DateCursor.GetDateFormat())
-		return
+	if err := requester.processHTTPRequest(stdCtx, publisher); err != nil {
+		return err
```
returning here will shut down the input. Is the error that critical that the input is eventually broken or can we recover (e.g. just log and try again after some backoff)?

urso commented on x-pack/filebeat/input/httpjson/input.go:
```diff
@@ -81,532 +48,145 @@ func newRetryLogger() *retryLogger {
-		return in.config.URL
-	}
+// Run starts the input worker then returns. Only the first invocation
+// will ever start the worker.
```
This comment seems to be wrong.

urso commented on x-pack/filebeat/input/httpjson/input.go:
```diff
@@ -81,532 +48,145 @@ func newRetryLogger() *retryLogger {
+	stdCtx := ctxtool.FromCanceller(ctx.Cancelation)
 
-	url, err := url.Parse(in.config.URL)
+	httpClient, err := in.newHTTPClient(stdCtx)
```
how can this fail? Does the client make a request to init oauth support? Are there recoverable error types (e.g. io timeout)? In the later case we might want to put the initialization into a loop with exponential backoff.

urso commented on x-pack/filebeat/input/httpjson/input.go:
```diff
@@ -81,532 +48,145 @@ func newRetryLogger() *retryLogger {
-		in.log.Warn("date_cursor field must be a string, cursor will not advance")
-		return
+	if in.config.Interval == 0 {
+		return nil
```
Is Interval==0 a configuration error? Maybe this is exactly what some users want, (long-)poll the server.

urso commented on x-pack/filebeat/input/httpjson/input.go:
```diff
@@ -81,532 +48,145 @@ func newRetryLogger() *retryLogger {
-				}
-			}
-		}
+	switch errors.Cause(err) {
```
Cause is used by some legacy error libs (e.g. pkg/errors which is not maintained anymore). Better use `Unpack` or `errors.Is(context.X)`.

urso commented on x-pack/filebeat/input/httpjson/input.go:
```diff
@@ -81,532 +48,145 @@ func newRetryLogger() *retryLogger {
-			}
-		}
+	switch errors.Cause(err) {
+	case context.Canceled, context.DeadlineExceeded:
```
is DeadlineExceeded a signal that might be triggered internally by processHTTPRequest, or do we expect this to be triggered via the inputs parent context?

urso commented on x-pack/filebeat/input/httpjson/requester.go:
```diff
@@ -0,0 +1,265 @@
+				}
+				msg, err := r.client.Do(req)
+				if err != nil {
+					return nil, errors.Wrapf(err, "failed to execute http client.Do")
```
It looks like this error is passed through to `Run`. If so the input shuts down after an IO error. Timeouts/disconnects or other network errors should not shut down the input. These are recoverable errors in a sense of "let's try again".

errors.Wrap and similar make it difficult to "judge" the error (`errors.Is` can help to check for a cause). Consider to either log recoverable errors or create special error types that allow you to see what's going on.

About `errors.Wrap`: We should stop using the pkg/errors package. It is discontinued. At the same time go 1.14 introduced something similar via `fmt.Errorf` with `%w`. Better use `fmt.Errorf("failed to execute http request: %w", err)`

urso commented on x-pack/filebeat/input/httpjson/requester.go:
```diff
@@ -0,0 +1,265 @@
+
+		responseData, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return errors.Wrapf(err, "failed to read http response")
```
IO error here should not force the input to shutdown.
